### ISSUE TITLE
Get error when creating volume snapshot

### ISSUE THREAD

**Issue by igorgonibm** [kubernetes-csi/external-snapshotter#300]
When creating snapshot I get the following messages in csi-snapshotter logs:
Full log, snap content describe and relevant yamls attached

I0421 09:23:09.254741       1 snapshot_controller.go:606] setAnnVolumeSnapshotBeingCreated: set annotation [snapshot.storage.kubernetes.io/volumesnapshot-being-created:yes] on content [snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762].
I0421 09:23:09.259302       1 snapshot_controller.go:179] updateContentStatusWithEvent[snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762]
I0421 09:23:09.263085       1 snapshot_controller.go:200] updating VolumeSnapshotContent[snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762] error status failed Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io "snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762": the object has been modified; please apply your changes to the latest version and try again
E0421 09:23:09.263110       1 snapshot_controller.go:139] createSnapshot [create-snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762]: error occurred in createSnapshotWrapper: failed to add VolumeSnapshotBeingCreated annotation on the content snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762: "snapshot controller failed to update snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762 on API server: Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io \"snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762\": the object has been modified; please apply your changes to the latest version and try again"
E0421 09:23:09.263208       1 goroutinemap.go:150] Operation for "create-snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762" failed. No retries permitted until 2020-04-21 09:23:09.763165038 +0000 UTC m=+6.819361135 (durationBeforeRetry 500ms). Error: "failed to add VolumeSnapshotBeingCreated annotation on the content snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762: \"snapshot controller failed to update snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762 on API server: Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io \\\"snapcontent-3f3c2bc8-2305-4b26-95f6-0e641e8e3762\\\": the object has been modified; please apply your changes to the latest version and try again\""
[yamls_and_log.zip](https://github.com/kubernetes-csi/external-snapshotter/files/4509035/yamls_and_log.zip)

**Comment by xing-yang** [comment]
I see the following error message in your logs.  It looks like you have not created the secret but you have that secret referenced in the snapshot class.

```
E0421 09:29:03.988544       1 snapshot_controller.go:535] Failed to get credentials for snapshot snapcontent-c62126d0-15af-4c45-a950-99e056a12604: error getting secret secret-67a in namespace kube-system: secrets "secret-67a" not found
```

**Comment by igorgonibm** [comment]
@xing-yang any idea what could be wrong?

**Comment by xing-yang** [comment]
@igorgonibm what image versions of snapshot-controller and csi-snapshotter sidecar are you using?

Can you provide "kubectl describe volumesnapshot"?   Those error messages you posted are failures updating the APIs but the retried were successful later as VolumeSnapshotContent has its status updated and ReadyToUse is set true.  I don't know what is the status of VolumeSnapshot.  Can you set log level to 5 and reproduce the logs?  I can't find any messages on updating the VoumeSnapshot status in the existing logs.

**Comment by igorgonibm** [comment]
I created new snapshot and attached relevant yamls. describe of class, snapshot and snapshotcomntent and snapshotter log
thank y
[logs.zip](https://github.com/kubernetes-csi/external-snapshotter/files/4516670/logs.zip)
ou

**Comment by xing-yang** [comment]
According to "kubectl describe volumesnapshot", the snapshot is created successfully.

```
Name:         snap20
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  snapshot.storage.k8s.io/v1beta1
Kind:         VolumeSnapshot
Metadata:
  Creation Timestamp:  2020-04-22T14:07:52Z
  Finalizers:
    snapshot.storage.kubernetes.io/volumesnapshot-as-source-protection
    snapshot.storage.kubernetes.io/volumesnapshot-bound-protection
  Generation:        1
  Resource Version:  2052306
  Self Link:         /apis/snapshot.storage.k8s.io/v1beta1/namespaces/default/volumesnapshots/snap20
  UID:               b52df980-4158-4af1-b0ad-bffa7d1a2937
Spec:
  Source:
    Persistent Volume Claim Name:  pvc-snapshot
  Volume Snapshot Class Name:      snapclass
Status:
  Bound Volume Snapshot Content Name:  snapcontent-b52df980-4158-4af1-b0ad-bffa7d1a2937
  Creation Time:                       2020-04-22T14:08:03Z
  Ready To Use:                        true
  Restore Size:                        1Gi
Events:                                <none>
```

`Ready To Use` is true.  Creation time, restore size, and bound volume snapshot content name are all set in the VolumeSnapshot status.

**Comment by igorgonibm** [comment]
@xing-yang so why is there error message in  the log? It is related to the current snap content.

E0422 14:08:03.234487       1 snapshot_controller_base.go:250] could not sync content "snapcontent-b52df980-4158-4af1-b0ad-bffa7d1a2937": failed to remove VolumeSnapshotBeingCreated annotation from the content snapcontent-b52df980-4158-4af1-b0ad-bffa7d1a2937: "snapshot controller failed to update snapcontent-b52df980-4158-4af1-b0ad-bffa7d1a2937 on API server: Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io \"snapcontent-b52df980-4158-4af1-b0ad-bffa7d1a2937\": the object has been modified; please apply your changes to the latest version and try again"

**Comment by igorgonibm** [comment]
> This is because VolumeSnapshotBeingCreated annotation was already removed earlier. This syncContent call has an older version of the object so it failed to remove it again, but it is actually already removed.
> 
> I'll take a look to see how to avoid this error.

Thank you. Will wait for your answer.

**Comment by igorgonibm** [comment]
There seems to be additional problem which may be related
Our controller doesn't receive CreateVolume request

Describe on pvs specifies that:
`failed to provision volume with StorageClass "sc-snapshot": error getting handle for DataSource Type VolumeSnapshot by Name snap20: error getting snapshot snap20 from api server: the server could not find the requested resource (get volumesnapshots.snapshot.storage.k8s.io snap20)
`
But according to this, the source exists.
[logs_pvc_from_snap.zip](https://github.com/kubernetes-csi/external-snapshotter/files/4523315/logs_pvc_from_snap.zip)

```

 kubectl get volumesnapshots.snapshot.storage.k8s.io snap20
NAME     READYTOUSE   SOURCEPVC      SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS   SNAPSHOTCONTENT                                    CREATIONTIME   AGE
snap20   true         pvc-snapshot                           1Gi           snapclass       snapcontent-b52df980-4158-4af1-b0ad-bffa7d1a2937   24h            24h

```
[logs_pvc_from_snap.zip](https://github.com/kubernetes-csi/external-snapshotter/files/4523321/logs_pvc_from_snap.zip)

**Comment by bhargavkeshav** [comment]
Hi @xing-yang,

I am facing similar issue , was this issue fixed.

```
E0927 04:04:34.862225       1 snapshot_controller_base.go:405] could not sync content "snapcontent-1d5d9c1a-8824-4087-b2a3-a824d00fb60c": snapshot controller failed to update snapcontent-1d5d9c1a-8824-4087-b2a3-a824d00fb60c on API server: Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io "snapcontent-1d5d9c1a-8824-4087-b2a3-a824d00fb60c": the object has been modified; please apply your changes to the latest version and try again
I0927 04:04:34.862272       1 snapshot_controller_base.go:272] Failed to sync content "snapcontent-1d5d9c1a-8824-4087-b2a3-a824d00fb60c", will retry again: snapshot controller failed to update snapcontent-1d5d9c1a-8824-4087-b2a3-a824d00fb60c on API server: Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io "snapcontent-1d5d9c1a-8824-4087-b2a3-a824d00fb60c": the object has been modified; please apply your changes to the latest version and try again
E0927 04:04:35.857893       1 snapshot_controller_base.go:379] could not sync snapshot "mission-control/velero-mission-control-data-mission-control-2-ztc2t": snapshot controller failed to update velero-mission-control-data-mission-control-2-ztc2t on API server: Operation cannot be fulfilled on volumesnapshots.snapshot.storage.k8s.io "velero-mission-control-data-mission-control-2-ztc2t": StorageError: invalid object, Code: 4, Key: /registry/snapshot.storage.k8s.io/volumesnapshots/mission-control/velero-mission-control-data-mission-control-2-ztc2t, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 42725590-e58f-40a0-b3c2-5b84f4dff2a0, UID in object meta: 
I0927 04:04:35.857944       1 snapshot_controller_base.go:197] Failed to sync snapshot "mission-control/velero-mission-control-data-mission-control-2-ztc2t", will retry again: snapshot controller failed to update velero-mission-control-data-mission-control-2-ztc2t on API server: Operation cannot be fulfilled on volumesnapshots.snapshot.storage.k8s.io "velero-mission-control-data-mission-control-2-ztc2t": StorageError: invalid object, Code: 4, Key: /registry/snapshot.storage.k8s.io/volumesnapshots/mission-control/velero-mission-control-data-mission-control-2-ztc2t, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 42725590-e58f-40a0-b3c2-5b84f4dff2a0, UID in object meta: 
E0927 04:04:36.056089       1 snapshot_controller_base.go:379] could not sync snapshot "distribution/velero-redis-data-distribution-1-w76fr": snapshot controller failed to update velero-redis-data-distribution-1-w76fr on API server: Operation cannot be fulfilled on volumesnapshots.snapshot.storage.k8s.io "velero-redis-data-distribution-1-w76fr": StorageError: invalid object, Code: 4, Key: /registry/snapshot.storage.k8s.io/volumesnapshots/distribution/velero-redis-data-distribution-1-w76fr, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: b9ce0d3a-f1c8-448d-a40c-a1ba6a1366c1, UID in object meta: 
I0927 04:04:36.056139       1 snapshot_controller_base.go:197] Failed to sync snapshot "distribution/velero-redis-data-distribution-1-w76fr", will retry again: snapshot controller failed to update velero-redis-data-distribution-1-w76fr on API server: Operation cannot be fulfilled on volumesnapshots.snapshot.storage.k8s.io "velero-redis-data-distribution-1-w76fr": StorageError: invalid object, Code: 4, Key: /registry/snapshot.storage.k8s.io/volumesnapshots/distribution/velero-redis-data-distribution-1-w76fr, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: b9ce0d3a-f1c8-448d-a40c-a1ba6a1366c1, UID in object meta: 
E0927 04:04:36.455871       1 snapshot_controller_base.go:379] could not sync snapshot "artifactory-ha/velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh": snapshot controller failed to update velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh on API server: Operation cannot be fulfilled on volumesnapshots.snapshot.storage.k8s.io "velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh": StorageError: invalid object, Code: 4, Key: /registry/snapshot.storage.k8s.io/volumesnapshots/artifactory-ha/velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 1d5d9c1a-8824-4087-b2a3-a824d00fb60c, UID in object meta: 
I0927 04:04:36.455925       1 snapshot_controller_base.go:197] Failed to sync snapshot "artifactory-ha/velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh", will retry again: snapshot controller failed to update velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh on API server: Operation cannot be fulfilled on volumesnapshots.snapshot.storage.k8s.io "velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh": StorageError: invalid object, Code: 4, Key: /registry/snapshot.storage.k8s.io/volumesnapshots/artifactory-ha/velero-volume-artifactory-ha-artifactory-ha-primary-1-2mlsh, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 1d5d9c1a-8824-4087-b2a3-a824d00fb60c, UID in object meta:
```

**Comment by Ahmad-Faizan** [comment]
this seems to be related issue https://github.com/kubernetes-sigs/controller-runtime/issues/1881

**Comment by RamyAllam** [comment]
I can see the same issue as well on GKE v1.25.8-gke.1000
```
# kubectl get volumesnapshot/test01-cqlzf -n mysql01-db-8moti --cluster gke_CLUSTER_CONTEXT -o yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  creationTimestamp: "2023-07-18T15:37:44Z"
  finalizers:
  - snapshot.storage.kubernetes.io/volumesnapshot-as-source-protection
  - snapshot.storage.kubernetes.io/volumesnapshot-bound-protection
  generation: 1
  name: test01-cqlzf
  namespace: mysql01-db-8moti
  resourceVersion: "225004131"
  uid: bd15ce02-8aef-4c3c-8d45-0acb92a5800c
spec:
  source:
    persistentVolumeClaimName: data-mysql01-db-8moti-mysql-0
  volumeSnapshotClassName: pd-snapshot
status:
  boundVolumeSnapshotContentName: snapcontent-bd15ce02-8aef-4c3c-8d45-0acb92a5800c
  creationTime: "2023-07-18T15:37:46Z"
  readyToUse: true
  restoreSize: 5Gi
```

Error
```
Failed to create snapshot: failed to add VolumeSnapshotBeingCreated annotation on the content snapcontent-bd15ce02-8aef-4c3c-8d45-0acb92a5800c: "snapshot controller failed to update snapcontent-bd15ce02-8aef-4c3c-8d45-0acb92a5800c on API server: Operation cannot be fulfilled on volumesnapshotcontents.snapshot.storage.k8s.io \"snapcontent-bd15ce02-8aef-4c3c-8d45-0acb92a5800c\": the object has been modified; please apply your changes to the latest version and try again"
```

**Comment by luckymrwang** [comment]
Hi @xing-yang 

I am facing similar issue , was this issue fixed ?

**Comment by xing-yang** [comment]
There's a WIP PR that may fix this problem: https://github.com/kubernetes-csi/external-snapshotter/pull/876

**Comment by luckymrwang** [comment]
> There's a WIP PR that may fix this problem: #876

ok
